opentelemetry-cpp: version update and extra configurability

universenox opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Update package versions:
    opentelemetry-cpp: update versions.
opentelemetry-proto 1.3.2 -> 1.5.0 https://github.com/open-telemetry/opentelemetry-proto/compare/v1.3.2...v1.5.0
opentelemetry 1.16.1 -> 1.20.0 https://github.com/open-telemetry/opentelemetry-cpp/compare/v1.16.1...v1.20.0

Note that the update script isn't quite right because there's actually two sources -- it (`nix-shell maintainers/scripts/update.nix --argstr package opentelemetry-cpp`) will update only opentelemetry-cpp but not the opentelemetry-proto. 

Add STL option. I think it's nicer to use, as a developer using the library (versus otlp-cpp's own stl-like implementations). Plus, using nix, I should be assured of retaining ABI compatibility, which appears to be the main reason not to use the STL here. Also I was surprised that I needed both WITH_STL and CMAKE_CXX_STANDARD cmake flags, versus just WITH_STL.

Make the exporter-specific dependencies opt-in since I don't want to pull in unnecessary ones.

Moved some packages to propagatedBuildInputs along the above change, since whoever's enabled the particular exporter is probably going to want to build with those dependencies. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

universenox committed:
opentelemetry-cpp: add option to use STL

universenox committed:
opentelemetry-cpp: increased configurability of cmake options and dependencies

Corristo commented:
Just a heads-up: opentelemetry-cpp released version 1.20.0 last month, so this is outdated already.

panicgh commented on pkgs/by-name/op/opentelemetry-cpp/package.nix:
```diff
@@ -53,17 +67,22 @@ stdenv.mkDerivation (finalAttrs: {
+  cmakeFlags =
+    [
+      (lib.cmakeBool "BUILD_SHARED_LIBS" (!stdenv.hostPlatform.isStatic))
+      (lib.cmakeBool "WITH_ABSEIL" true)
```
`WITH_ABSEIL` is removed in the commit after the 1.20.0 release. So probably abseil-cpp and the configure flag should be removed from here, too, since they are not future proof. See https://github.com/open-telemetry/opentelemetry-cpp/issues/3116

universenox committed:
opentelemetry-cpp: update versions.
opentelemetry-proto 1.3.2 -> 1.5.0 https://github.com/open-telemetry/opentelemetry-proto/compare/v1.3.2...v1.5.0
opentelemetry 1.16.1 -> 1.20.0 https://github.com/open-telemetry/opentelemetry-cpp/compare/v1.16.1...v1.20.0

Co-authored-by: panicgh <79252025+panicgh@users.noreply.github.com>

panicgh commented on pkgs/by-name/op/opentelemetry-cpp/package.nix:
```diff
@@ -53,17 +66,21 @@ stdenv.mkDerivation (finalAttrs: {
+      (lib.cmakeFeature "OTELCPP_PROTO_PATH" "${opentelemetry-proto}")
+    ]
+    ++ lib.optionals (cxxStandard != null) [
+      (lib.cmakeFeature "CMAKE_CXX_STANDARD" cxxStandard)
```
I will not insist on removing CMAKE_CXX_STANDARD. I think having it does no harm.

panicgh commented on pkgs/by-name/op/opentelemetry-cpp/package.nix:
```diff
@@ -53,17 +66,21 @@ stdenv.mkDerivation (finalAttrs: {
+    ]
+    ++ lib.optionals (cxxStandard != null) [
+      (lib.cmakeFeature "CMAKE_CXX_STANDARD" cxxStandard)
+      (lib.cmakeFeature "WITH_STL" "CXX${cxxStandard}")
```
Thanks for the clarification. I think a separate is not necessary.

panicgh commented on pkgs/by-name/op/opentelemetry-cpp/package.nix:
```diff
@@ -48,9 +47,7 @@ stdenv.mkDerivation (finalAttrs: {
-  propagatedBuildInputs =
-    [ abseil-cpp ]
-    ++ lib.optionals enableGrpc [
+  propagatedBuildInputs = lib.optionals enableGrpc [
```
There is a nixfmt error (which I introduced via a previous suggestion). Otherwise it looks good to me.

Corristo commented on pkgs/by-name/op/opentelemetry-cpp/package.nix:
```diff
@@ -39,12 +44,18 @@ stdenv.mkDerivation (finalAttrs: {
 
+  propagatedBuildInputs =
+    lib.optionals enableGrpc [
+      protobuf
```
protobuf is not only required by the grpc exporter, but also by the http exporter. The condition for `lib.optionals` therefore should be `(enableGrpc || enableHttp)`.

panicgh reviewed (approved):
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 390485`
Commit: `58f9fdb6925d0247c86b8beed73cb57929e5b88f`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>opentelemetry-cpp</li>
    <li>opentelemetry-cpp.dev</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5668

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2494
